File: api-java/src/org/glite/rgma/Registry.java
/*
 * Copyright (c) Members of the EGEE Collaboration. 2004.
 * See http://eu-egee.org/partners/ for details on the copyright holders.
 * For license conditions see the license file or http://eu-egee.org/license.html
 */
package org.glite.rgma;

import java.util.ArrayList;
import java.util.List;

/**
 * A registry provides information about available producers.
 */
public class Registry extends VDB {

	/**
	 * Constructs a registry object for the specified VDB.
	 * 
	 * @param vdbName
	 *            name of VDB
	 */
	public Registry(String vdbName) {
		m_servletName = "RegistryServlet";
		m_vdbName = vdbName;
	}

	/**
	 * Returns a list of all producers for a table with their registered information.
	 * 
	 * @param name
	 *            name of table
	 * 
	 * @return a list of producer table entry objects, one for each producer
	 * @throws RGMAPermanentException
	 * @throws RGMATemporaryException
	 */
	public List<ProducerTableEntry> getAllProducersForTable(String name) throws RGMAPermanentException, RGMATemporaryException {
		ServletConnection connection = getNewConnection();
		connection.addParameter("canForward", true);
		connection.addParameter("tableName", name);
		ArrayList<ProducerTableEntry> pdl = new ArrayList<ProducerTableEntry>();
		for (Tuple t : connection.sendNonResourceCommand("getAllProducersForTable").getData()) {
			ResourceEndpoint endpoint = new ResourceEndpoint(t.getString(0), t.getInt(1));
			ProducerTableEntry pd = new ProducerTableEntry(endpoint, t.getBoolean(2), t
					.getBoolean(3), t.getBoolean(4), t.getBoolean(5), t.getBoolean(6), t.getString(7), t.getInt(8));
			pdl.add(pd);
		}
		return pdl;
	}
}

File: api-cpp/src/Registry.cpp
#include "rgma/Registry.h"
#include "rgma/TupleSet.h"
#include "rgma/URLBuilder.h"
#include "rgma/ResourceEndpoint.h"

namespace glite {
namespace rgma {

Registry::Registry(const std::string & vdbName) throw(RGMATemporaryException, RGMAPermanentException) :
    m_vdbName(vdbName), m_connection(URLBuilder::getURL("Registry")) {
}

void Registry::clearServletConnection() {
    m_connection.clear();
    m_connection.addParameter("vdbName", m_vdbName);
}

std::vector<ProducerTableEntry> Registry::getAllProducersForTable(const std::string & tableName)
        throw(RGMATemporaryException, RGMAPermanentException) {
    clearServletConnection();
    m_connection.addParameter("tableName", tableName);
    m_connection.addParameter("canForward", true);
    TupleSet result;
    m_connection.connect("getAllProducersForTable", result);
    TupleSet::const_iterator it(result.begin());
    std::vector<ProducerTableEntry> pdl;
    while (it != result.end()) {
        ResourceEndpoint endpoint(it->getString(0), it->getInt(1));
        ProducerTableEntry pd(endpoint, it->getBool(2), it->getBool(3), it->getBool(4), it->getBool(5), it->getBool(6),
                it->getString(7), it->getInt(8));
        pdl.push_back(pd);
        ++it;
    }
    return pdl;
}

}
}
